api/src/main/java/org/openmrs/api/context/Daemon.java

```diff
@@ -390,30 +360,114 @@ public void run() {
 				}
 			}
 		};
-		
-		thread.start();
+
+		OpenmrsThreadPoolHolder.threadExecutor.execute(thread);
 		return thread;
 	}
+
+	/**
+	 * Executes the given runnable in a new thread that is authenticated as the daemon user.
+	 *
+	 * @param callable an object implementing the {@link Callable<T>} interface to be run
+	 * @param token the token required to run code as the daemon user
+	 * @return the newly spawned {@link Thread}
+	 * @since 2.7.0
+	 */
+	@SuppressWarnings({"squid:S1217", "unused"})
+	public static <T> Future<T> runInDaemonThread(final Callable<T> callable, DaemonToken token) {
+		if (!ModuleFactory.isTokenValid(token)) {
+			throw new ContextAuthenticationException("Invalid token");
+		}
+		
+		return runInDaemonThreadInternal(callable);
+	}
+
+	/**
+	 * Executes the given runnable in a new thread that is authenticated as the daemon user.
+	 *
+	 * @param runnable an object implementing the {@link Runnable} interface to be run
+	 * @param token the token required to run code as the daemon user
+	 * @return the newly spawned {@link Thread}
+	 * @since 2.7.0
+	 */
+	@SuppressWarnings("squid:S1217")
+	public static Future<?> runInDaemonThreadWithoutResult(final Runnable runnable, DaemonToken token) {
+		if (!ModuleFactory.isTokenValid(token)) {
+			throw new ContextAuthenticationException("Invalid token");
+		}
+
+		return runInDaemonThreadInternal(runnable);
+	}
 	
 	/**
 	 * Executes the given runnable in a new thread that is authenticated as the daemon user and wait
 	 * for the thread to finish.
 	 *
 	 * @param runnable an object implementing the {@link Runnable} interface.
 	 * @param token the token required to run code as the daemon user
-	 * @since 1.9.2
+	 * @since 2.7.0
 	 */
 	public static void runInDaemonThreadAndWait(final Runnable runnable, DaemonToken token) {
-		Thread daemonThread = runInDaemonThread(runnable, token);
+		Future<?> daemonThread = runInDaemonThreadWithoutResult(runnable, token);
 		
 		try {
-			daemonThread.join();
+			daemonThread.get();
 		}
-		catch (InterruptedException e) {
-			//Ignore
+		catch (InterruptedException | ExecutionException e) {
+			// Ignored
 		}
 	}
 	
+	private static <T> Future<T> runInDaemonThreadInternal(Callable<T> callable) {
+		return OpenmrsThreadPoolHolder.threadExecutor.submit(() -> {
+			isDaemonThread.set(true);
+			try {
+				Context.openSession();
+				return callable.call();
+			}
+			finally {
+				try {
+					Context.closeSession();
+				} finally {
+					isDaemonThread.remove();
+				}
+			}
+		});
+	}
+	
+	private static Future<?> runInDaemonThreadInternal(Runnable runnable) {
+		// for Threads, we used to guarantee that Thread.start() was called before the function returned
+		// since we cannot guarantee that the executor actually started executing the thread, we use a CountDownLatch
+		// to emulate this behaviour when the user submits a Thread. Other runnables are unaffected.
+		CountDownLatch countDownLatch = getCountDownLatch(runnable instanceof Thread);
```

Comment:
There are two differences between this function and the one that operates on `Callable`s: the obvious difference that a `Runnable` has no result, and the more subtle distinction that a `Thread` is a `Runnable`. Any `Thread` objects that are submitted will be guaranteed to resume _after_ the thread has been started and the session is open.

The reason for the above hack is to continue to support this idiom:

```java
Thread myThread = new Thread() {
	// snip
}

runInDaemonThread(myThread, token);

thread.join();
```

Basically if `runInDaemonThread()` returns control to consume code before the thread is actually started, it's easy to get in a state where the thread.join()` call returns immediately because the thread isn't running yet. The delay until we hit the `CountDownLatch` ensures that when `thread.join()` is called the thread will have been started (though, obviously we make no guarantee that the thread has not finished by the time it's returned.